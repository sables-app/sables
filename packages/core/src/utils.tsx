import { createMutableRef, MutableReferenceObjectValue } from "@sables/utils";

import type * as ReduxToolkit from "@reduxjs/toolkit";
import type { ComponentProps, ElementType, FunctionComponent } from "react";
import type * as Redux from "redux";

import {
  LIFECYCLE_REF_MESSAGE,
  ROUTES_COLLECTION_REF_MESSAGE,
  SYMBOL_EFFECT_API_LIFECYCLE,
  SYMBOL_EFFECT_API_ROUTES,
  SYMBOL_LAZY_META,
  SYMBOL_LAZY_SELECTOR,
} from "./constants.js";
import { ActionSubject } from "./main.js";
import type {
  AnyLazySelector,
  DefaultEffectAPI,
  LazyMeta,
  ObjectWithLazyMeta,
} from "./types.js";

/** @internal */
export function hasLazyMeta<T extends object>(
  value: T
): value is T & ObjectWithLazyMeta {
  return Object.hasOwn(value, SYMBOL_LAZY_META);
}

/** @internal */
export function createLazyMeta(): LazyMeta {
  return {
    observers: new Set(),
    slices: new Set(),
  };
}

/** @internal */
export function isEnhancedSelector(value: unknown): value is AnyLazySelector {
  return (
    typeof value == "function" && Object.hasOwn(value, SYMBOL_LAZY_SELECTOR)
  );
}

/** @internal */
export function getSlicesFromSelector(
  selector: ReduxToolkit.Selector
): ReduxToolkit.Slice[] {
  return isEnhancedSelector(selector)
    ? [...selector[SYMBOL_LAZY_META].slices]
    : [];
}

type LifecycleRef<
  StoreState extends Record<string, unknown> = Record<string, unknown>
> = DefaultEffectAPI<StoreState>[typeof SYMBOL_EFFECT_API_LIFECYCLE];

type RoutesCollectionRef<
  StoreState extends Record<string, unknown> = Record<string, unknown>
> = DefaultEffectAPI<StoreState>[typeof SYMBOL_EFFECT_API_ROUTES];

/** @internal */
export function createEffectAPIDefaults<
  StoreState extends Record<string, unknown> = Record<string, unknown>
>({
  actions$,
  dispatch,
  getState,
  lifecycleRef,
  routesCollectionRef,
}: {
  actions$?: ActionSubject;
  dispatch: ReduxToolkit.Dispatch;
  getState: Redux.Store<StoreState>["getState"];
  lifecycleRef?: LifecycleRef<StoreState>;
  routesCollectionRef?: RoutesCollectionRef<StoreState>;
}): DefaultEffectAPI<StoreState> {
  return {
    actions$: actions$ || new ActionSubject(),
    dispatch,
    getState,
    [SYMBOL_EFFECT_API_LIFECYCLE]:
      lifecycleRef ||
      createMutableRef<MutableReferenceObjectValue<LifecycleRef>>(
        LIFECYCLE_REF_MESSAGE
      ),
    [SYMBOL_EFFECT_API_ROUTES]:
      routesCollectionRef ||
      createMutableRef<MutableReferenceObjectValue<RoutesCollectionRef>>(
        ROUTES_COLLECTION_REF_MESSAGE
      ),
  };
}

/**
 * A higher-order component that assigns props to the given component.
 *
 * @see {@link https://sables.dev/api#withprops withProps documentation}
 *
 * @example
 *
 * const FooLink = withProps("a", { className: "foo" });
 * const ButtonFooLink = withProps(LinkWithClassName, { role: "button" });
 *
 * @public
 */
export function withProps<C extends ElementType>(
  Component: C,
  baseProps: ComponentProps<C>
): FunctionComponent<ComponentProps<C>> {
  function ComponentWithProps(props: ComponentProps<C>) {
    return <Component {...baseProps} {...props} />;
  }

  const originalDisplayName =
    (typeof Component == "function" && Component.displayName) || "Component";

  ComponentWithProps.displayName = `withProps(${originalDisplayName})`;

  return ComponentWithProps;
}
