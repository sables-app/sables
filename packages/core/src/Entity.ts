import { capitalize } from "@sables/utils";

import type * as ReduxToolkit from "@reduxjs/toolkit";

import { createSelector } from "./Selector.js";
import { PayloadAction } from "./types.js";

/** @internal */
type ExtractEntityFromAdapter<T> = T extends ReduxToolkit.EntityStateAdapter<
  infer Entity
>
  ? Entity
  : never;

/** @internal */
type ExtractEntityFromSelector<T> = T extends ReduxToolkit.EntitySelectors<
  infer Entity,
  any
>
  ? Entity
  : never;

/** @internal */
export namespace EntityReducers {
  // prettier-ignore
  export type ReducerName<
    K,
    SingularName extends string,
    PluralName extends string
  > = K extends "addOne"     ? `add${Capitalize<SingularName>}`     : never
    | K extends "addMany"    ? `add${Capitalize<PluralName>}`       : never
    | K extends "setOne"     ? `set${Capitalize<SingularName>}`     : never
    | K extends "setMany"    ? `set${Capitalize<PluralName>}`       : never
    | K extends "setAll"     ? `setAll${Capitalize<PluralName>}`    : never
    | K extends "removeOne"  ? `remove${Capitalize<SingularName>}`  : never
    | K extends "removeMany" ? `remove${Capitalize<PluralName>}`    : never
    | K extends "removeAll"  ? `removeAll${Capitalize<PluralName>}` : never
    | K extends "updateOne"  ? `update${Capitalize<SingularName>}`  : never
    | K extends "updateMany" ? `update${Capitalize<PluralName>}`    : never
    | K extends "upsertOne"  ? `upsert${Capitalize<SingularName>}`  : never
    | K extends "upsertMany" ? `upsert${Capitalize<PluralName>}`    : never;

  export type Reducers<
    Entity,
    SingularName extends string,
    PluralName extends string,
  > = {
    [K in keyof ReduxToolkit.EntityStateAdapter<Entity> as ReducerName<
      K,
      SingularName,
      PluralName
    >]: ReduxToolkit.EntityStateAdapter<Entity>[K];
  };
}

/** @internal */
export namespace EntitySelectors {
  // prettier-ignore
  export type SelectorName<
    K,
    SingularName extends string,
    PluralName extends string
  > =
    | K extends "selectAll"      ? `selectAll${Capitalize<PluralName>}`          : never
    | K extends "selectById"     ? `select${Capitalize<SingularName>}ById`       : never
    | K extends "selectTotal"    ? `select${Capitalize<SingularName>}Count`      : never
    | K extends "selectEntities" ? `select${Capitalize<SingularName>}Dictionary` : never
    | K extends "selectIds"      ? `select${Capitalize<SingularName>}Ids`        : never;

  export type Selectors<
    Entity,
    SingularName extends string,
    PluralName extends string,
  > = {
    [K in keyof ReduxToolkit.EntitySelectors<
      Entity,
      Record<string, unknown>
    > as SelectorName<
      K,
      SingularName,
      PluralName
    >]: ReduxToolkit.EntitySelectors<Entity, Record<string, unknown>>[K];
  };
}

/**
 * Assigns distinctive names to entity adapter case reducers.
 *
 * @see {@link https://sables.dev/api#distinctentityreducers `distinctEntityReducers` documentation}
 *
 * @example
 *
 * interface Dog {
 *   id: number;
 *   name: string;
 * }
 *
 * const dogsEntityAdapter = createEntityAdapter<Dog>();
 *
 * const dogsSlice = createSlice(
 *   "dogs",
 *   dogsEntityAdapter.getInitialState()
 * ).setReducer((builder) =>
 *   builder.addCases(
 *     distinctEntityReducers(dogsEntityAdapter, "dog")
 *   )
 * );
 *
 * // `addDog` is equal to `addOne` from Redux Toolkit
 * const { addDog } = dogsSlice.actions;
 *
 * @public
 */
export function distinctEntityReducers<
  Adapter extends ReduxToolkit.EntityStateAdapter<any>,
  SingularName extends string,
  PluralName extends string = `${SingularName}s`,
>(
  adapter: Adapter,
  name: SingularName | [SingularName, PluralName],
): EntityReducers.Reducers<
  ExtractEntityFromAdapter<Adapter>,
  SingularName,
  PluralName
> {
  const [singular, plural] = typeof name === "string" ? [name] : name;
  const singularName = capitalize(singular);
  const pluralName = capitalize(plural ?? `${singular}s`);

  return {
    [`add${singularName}`]: adapter.addOne,
    [`add${pluralName}`]: adapter.addMany,
    [`set${singularName}`]: adapter.setOne,
    [`set${pluralName}`]: adapter.setMany,
    [`setAll${pluralName}`]: adapter.setAll,
    [`remove${singularName}`]: adapter.removeOne,
    [`remove${pluralName}`]: adapter.removeMany,
    [`removeAll${pluralName}`]: adapter.removeAll,
    [`update${singularName}`]: adapter.updateOne,
    [`update${pluralName}`]: adapter.updateMany,
    [`upsert${singularName}`]: adapter.upsertOne,
    [`upsert${pluralName}`]: adapter.upsertMany,
  } as any;
}

/**
 * Assigns distinctive names to entity adapter selectors.
 *
 * @see {@link https://sables.dev/api#distinctentityselectors `distinctEntitySelectors` documentation}
 *
 * @example
 *
 * interface Book {
 *   id: number;
 *   name: string;
 * }
 *
 * const booksEntityAdapter = createEntityAdapter<Book>();
 * const booksSlice = createSlice(
 *   "books",
 *   booksEntityAdapter.getInitialState()
 * ).setReducer((builder) => builder);
 *
 * // `selectBookCount` is equal to `selectTotal` from Redux Toolkit
 * const { selectBookCount } = distinctEntitySelectors(
 *   booksEntityAdapter.getSelectors(booksSlice.selector),
 *   "book"
 * );
 *
 * @public
 */
export function distinctEntitySelectors<
  Selectors extends ReduxToolkit.EntitySelectors<any, any>,
  SingularName extends string,
  PluralName extends string = `${SingularName}s`,
>(
  selectors: Selectors,
  name: SingularName | [SingularName, PluralName],
): EntitySelectors.Selectors<
  ExtractEntityFromSelector<Selectors>,
  SingularName,
  PluralName
> {
  const [singular, plural] = typeof name === "string" ? [name] : name;
  const singularName = capitalize(singular);
  const pluralName = capitalize(plural ?? `${singular}s`);

  return {
    [`selectAll${pluralName}`]: selectors.selectAll,
    [`select${singularName}ById`]: selectors.selectById,
    [`select${singularName}Count`]: selectors.selectTotal,
    [`select${singularName}Dictionary`]: selectors.selectEntities,
    [`select${singularName}Ids`]: selectors.selectIds,
  } as any;
}

/** @internal */
export namespace NotableEntities {
  type AdjectiveEntityIDGetter<Payload> = {
    (actionPayload: Payload): ReduxToolkit.EntityId | undefined;
  };

  export type Adjectives = {
    [adjective: string]: AdjectiveEntityIDGetter<never>;
  };

  export type SliceStatePropertyName<
    Adjective extends string,
    Name extends string,
  > = `${Extract<Adjective, string>}${Capitalize<Name>}Payload`;

  export type PayloadFromAdjectiveFn<T extends AdjectiveEntityIDGetter<never>> =
    Parameters<T>[0];

  export type SliceState<
    Entity,
    A extends Adjectives,
    Name extends string,
  > = ReduxToolkit.EntityState<Entity> & {
    [K in keyof A as SliceStatePropertyName<
      Extract<K, string>,
      Name
    >]?: PayloadFromAdjectiveFn<A[K]>;
  };

  export type ReducerName<
    Adjective extends string,
    Name extends string,
  > = `denote${Capitalize<Extract<Adjective, string>>}${Capitalize<Name>}`;

  export type EntitySelectorName<
    TAdjective extends string,
    Name extends string,
  > = `select${Capitalize<Extract<TAdjective, string>>}${Capitalize<Name>}`;

  export type IdSelectorName<
    Adjective extends string,
    Name extends string,
  > = `select${Capitalize<Extract<Adjective, string>>}${Capitalize<Name>}Id`;

  export type PayloadSelectorName<
    Adjective extends string,
    Name extends string,
  > = `select${Capitalize<
    Extract<Adjective, string>
  >}${Capitalize<Name>}Payload`;

  export type Selectors<
    StoreState,
    Entity,
    A extends Adjectives,
    Name extends string,
  > = {
    [K in keyof A as EntitySelectorName<Extract<K, string>, Name>]: (
      storeState: StoreState,
    ) => Entity | undefined;
  } & {
    [K in keyof A as IdSelectorName<Extract<K, string>, Name>]: (
      storeState: StoreState,
    ) => ReduxToolkit.EntityId | undefined;
  } & {
    [K in keyof A as PayloadSelectorName<Extract<K, string>, Name>]: (
      storeState: StoreState,
    ) => NotableEntities.PayloadFromAdjectiveFn<A[K]> | undefined;
  };

  export type Instance<Entity, A extends Adjectives, Name extends string> = {
    getSelectors<StoreState>(
      selectState: (storeState: StoreState) => SliceState<Entity, A, Name>,
    ): Selectors<StoreState, Entity, A, Name>;
    getInitialState(): SliceState<Entity, A, Name>;
    reducers: {
      [K in keyof A as ReducerName<
        Extract<K, string>,
        Name
      >]: ReduxToolkit.CaseReducer<
        // NotableEntitiesState<Entity, A, Name> & Record<string, unknown>,
        //
        // TODO - Find a way to avoid using `any` here without forcing devs to
        // manually add types in generics. Using a type other than `any` results
        // in a type error.
        //
        // The only side effect of using `any` here is that reducers will accept
        // `any` slice state; all other types are maintained. This is only an issue
        // if the dev calls the reducer directly from outside of the slice.
        // Then the provided slice state won't be type checked. This isn't an expected
        // usage pattern, so this improvement is low priority.
        any,
        PayloadAction<NotableEntities.PayloadFromAdjectiveFn<A[K]>>
      >;
    };
  };
}

/**
 * Creates actions and reducers for notable entities.
 *
 * @see {@link https://sables.dev/api#createnotableentities `createNotableEntities` documentation}
 *
 * @public
 */
export function createNotableEntities<
  Adapter extends ReduxToolkit.EntityStateAdapter<any>,
  A extends NotableEntities.Adjectives,
  Name extends string,
>(
  _adapter: Adapter,
  singularName: Name,
  adjectives: A,
): NotableEntities.Instance<ExtractEntityFromAdapter<Adapter>, A, Name> {
  type State = NotableEntities.SliceState<
    ExtractEntityFromAdapter<Adapter>,
    A,
    Name
  >;

  function getStatePropertyName(
    adjective: string,
    name: string,
  ): NotableEntities.SliceStatePropertyName<typeof adjective, typeof name> {
    return `${adjective}${capitalize(name)}Payload`;
  }

  function getReducerName(
    adjective: string,
    name: string,
  ): NotableEntities.ReducerName<typeof adjective, typeof name> {
    return `denote${capitalize(adjective)}${capitalize(name)}`;
  }

  function getEntitySelectorName(
    adjective: string,
    name: string,
  ): NotableEntities.EntitySelectorName<typeof adjective, typeof name> {
    return `select${capitalize(adjective)}${capitalize(name)}`;
  }

  function getIdSelectorName(
    adjective: string,
    name: string,
  ): NotableEntities.IdSelectorName<typeof adjective, typeof name> {
    return `select${capitalize(adjective)}${capitalize(name)}Id`;
  }

  function getPayloadSelectorName(
    adjective: string,
    name: string,
  ): NotableEntities.PayloadSelectorName<typeof adjective, typeof name> {
    return `select${capitalize(adjective)}${capitalize(name)}Payload`;
  }

  function getSelectors(selectState: (storeState: any) => State) {
    return Object.entries(adjectives).reduce(
      (result, [adjective, getEntityId]) => {
        type Payload = NotableEntities.PayloadFromAdjectiveFn<
          typeof getEntityId
        >;
        const entitySelectorName = getEntitySelectorName(
          adjective,
          singularName,
        );
        const idSelectorName = getIdSelectorName(adjective, singularName);
        const payloadSelectorName = getPayloadSelectorName(
          adjective,
          singularName,
        );
        const statePropertyName = getStatePropertyName(adjective, singularName);
        const payloadSelector = createSelector(
          selectState,
          (state) => (state as any)[statePropertyName],
        );
        const idSelector = createSelector(payloadSelector, (payload) => {
          return payload !== undefined
            ? getEntityId(payload as Payload)
            : undefined;
        });
        const entitySelector = createSelector(
          selectState,
          idSelector,
          (sliceState, id) => {
            return id ? sliceState.entities[id] : undefined;
          },
        );

        return {
          ...result,
          [entitySelectorName]: entitySelector,
          [idSelectorName]: idSelector,
          [payloadSelectorName]: payloadSelector,
        };
      },
      {},
    );
  }

  function getInitialState() {
    return Object.keys(adjectives).reduce((result, adjective) => {
      const statePropertyName = getStatePropertyName(adjective, singularName);

      return {
        ...result,
        [statePropertyName]: undefined,
      };
    }, {});
  }

  const reducers = Object.entries(adjectives).reduce(
    (result, [adjective, _getEntityId]) => {
      const reducerName = getReducerName(adjective, singularName);
      const statePropertyName = getStatePropertyName(adjective, singularName);
      type Action = PayloadAction<
        NotableEntities.PayloadFromAdjectiveFn<typeof _getEntityId>
      >;

      return {
        ...result,
        [reducerName]: (state: Record<string, unknown>, action: Action) => {
          state[statePropertyName] = action.payload;
        },
      };
    },
    {},
  );

  return {
    getInitialState,
    getSelectors,
    reducers,
  } as any;
}
