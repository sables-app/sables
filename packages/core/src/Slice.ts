import { capitalize, demandValue } from "@sables/utils";

import type * as ReduxToolkit from "@reduxjs/toolkit";

import { createSliceReduxToolkit } from "../deps.js";
import { enhanceAction } from "./Action.js";
import { SYMBOL_LAZY_META, SYMBOL_LAZY_SELECTOR } from "./constants.js";
import type {
  AnySlice,
  BaseSliceLazySelector,
  EnhancedActionCreator,
  EnhancedSlice,
  LazySliceSelectorName,
  LazySliceStoreState,
  SliceLazySelector,
  SliceName,
  StandardActionCreator,
} from "./types.js";
import { createLazyMeta } from "./utils.js";

function createLazySliceSelectorName<S extends ReduxToolkit.Slice>(
  slice: S
): LazySliceSelectorName<S> {
  return `select${capitalize(slice.name)}State`;
}

function createLazySliceSelector<S extends ReduxToolkit.Slice>(
  enhancedSliceShell: Partial<EnhancedSlice<S>>
): SliceLazySelector<S> {
  const stateProperty: SliceName<S> = demandValue(enhancedSliceShell.name);
  const baseLazySelector: BaseSliceLazySelector<S> = function lazySelector(
    storeState?: LazySliceStoreState<S>
  ) {
    if (storeState === undefined || !Object.hasOwn(storeState, stateProperty)) {
      // The slice has not been added yet, so the initial state is used instead.
      return demandValue(enhancedSliceShell.getInitialState)();
    }

    return storeState[stateProperty];
  };

  const lazySelector: SliceLazySelector<S> = Object.assign(baseLazySelector, {
    [SYMBOL_LAZY_SELECTOR]: undefined,
    [SYMBOL_LAZY_META]: createLazyMeta(),
  });

  lazySelector[SYMBOL_LAZY_META].slices.add(
    enhancedSliceShell as unknown as ReduxToolkit.Slice
  );

  return lazySelector;
}

function enhanceSliceActions<S extends AnySlice>(
  actions: S["actions"],
  enhancedSliceShell: Partial<EnhancedSlice<S>>
): EnhancedSlice<S>["actions"] {
  type ActionCreatorEntries = Array<[string, StandardActionCreator<string>]>;
  type LazyActionCreatorEntries = Array<
    [string, EnhancedActionCreator<StandardActionCreator<string>>]
  >;
  const actionCreatorEntries = Object.entries(actions) as ActionCreatorEntries;
  const lazyActionCreatorEntries: LazyActionCreatorEntries =
    actionCreatorEntries.map(([name, actionCreator]) => {
      const nextActionCreator = enhanceAction(actionCreator);

      nextActionCreator[SYMBOL_LAZY_META].slices.add(
        enhancedSliceShell as ReduxToolkit.Slice
      );

      return [name, nextActionCreator];
    });
  const nextActions = Object.fromEntries(lazyActionCreatorEntries);

  return nextActions as any;
}

/**
 * Enhances the given slice by adding a named selector,
 * and actions that can be set as a slice dependent.
 *
 * @example
 *
 * const dogsSliceBasic = ReduxToolkit.createSlice({
 *   name: "dogs",
 *   initialState: {},
 *   reducers: {},
 * });
 *
 * const dogsSlice = enhanceSlice(dogsSliceBasic);
 *
 * dogsSlice.selector(store.getState());
 * dogsSlice.selectDogsState(store.getState());
 *
 * @public
 */
export function enhanceSlice<S extends ReduxToolkit.Slice>(
  slice: S
): EnhancedSlice<S> {
  const selectorName = createLazySliceSelectorName(slice);

  function isEnhancedSlice(
    value: ReduxToolkit.Slice | EnhancedSlice<S>
  ): value is EnhancedSlice<S> {
    return Object.hasOwn(value, selectorName);
  }

  if (isEnhancedSlice(slice)) return slice;

  const { actions, ...otherSliceProps } = slice;
  /**
   * A plain object that will be mutated to become the enhanced slice.
   * This object should be referenced as the returned enhanced slice.
   */
  const enhancedSliceShell = otherSliceProps as Partial<EnhancedSlice<S>>;
  const lazySelector = createLazySliceSelector(enhancedSliceShell);
  const enhancedActions = enhanceSliceActions(actions, enhancedSliceShell);

  return Object.assign(enhancedSliceShell, {
    ...otherSliceProps,
    [selectorName]: lazySelector,
    actions: enhancedActions,
    selector: lazySelector,
  }) as EnhancedSlice<S>;
}

/**
 * An enhanced version of Redux Toolkit's `createSlice`.
 *
 * @see {@link https://sables.dev/api#createslice `createSlice` documentation}
 * @see {@link https://redux-toolkit.js.org/api/createslice `ReduxToolkit.createSlice` documentation}
 * @see {@link enhanceSlice}
 *
 * @public
 */
export function createSlice<
  State,
  CaseReducers extends ReduxToolkit.SliceCaseReducers<State>,
  Name extends string = string
>(options: ReduxToolkit.CreateSliceOptions<State, CaseReducers, Name>) {
  return enhanceSlice(createSliceReduxToolkit(options));
}
