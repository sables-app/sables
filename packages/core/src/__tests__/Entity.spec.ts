import * as ReduxToolkit from "@reduxjs/toolkit";
import * as vitest from "vitest";
import { beforeAll, beforeEach, describe, expect, it, test } from "vitest";

import {
  combineReducers,
  configureStore,
  createEntityAdapter,
} from "../../deps.js";
import { createNotableEntities, entityAdapterToReducers } from "../Entity.js";
import { createSlice } from "../Slice.js";
import { createTestStore } from "./utils.js";

describe("Entity", () => {
  type Book = { bookId: string; title: string };

  let booksAdapter: ReduxToolkit.EntityAdapter<Book>;

  beforeAll(() => {
    booksAdapter = createEntityAdapter<Book>({
      selectId: (book) => book.bookId,
      sortComparer: (a, b) => a.title.localeCompare(b.title),
    });
  });

  describe("entityAdapterToReducers", () => {
    let bookReducers: ReturnType<typeof createTestConstructs>["bookReducers"];
    let booksSlice: ReturnType<typeof createTestConstructs>["booksSlice"];

    function createTestConstructs() {
      const bookReducers = entityAdapterToReducers(booksAdapter, "book");
      const booksSlice = createSlice(
        "books",
        booksAdapter.getInitialState()
      ).setReducer((builder) => builder.addCases(bookReducers));

      return { bookReducers, booksSlice };
    }

    beforeEach(() => {
      const testConstructs = createTestConstructs();

      bookReducers = testConstructs.bookReducers;
      booksSlice = testConstructs.booksSlice;
    });

    test("slice dependency integration", () => {
      const { store } = createTestStore(vitest);

      expect(store.getState()).toEqual(undefined);

      store.dispatch(
        booksSlice.actions.addBook({
          bookId: "1",
          title: "Harry Potter",
        })
      );

      expect(store.getState().books).toMatchSnapshot();

      expect(
        bookReducers.addBook(store.getState().books, {
          bookId: "1",
          title: "Harry Potter",
        })
      ).toMatchSnapshot();

      const booksSelectors = booksAdapter.getSelectors(
        booksSlice.selectBooksState
      );

      expect(booksSelectors.selectAll(store.getState())).toMatchSnapshot();
    });

    it("renames actions and reducers from entity adapters", () => {
      expect(
        booksSlice.actions.addBook({
          bookId: "3",
          title: "Goosebumps: Night of the Living Dummy",
        }).payload
      ).toMatchSnapshot();
      expect(
        booksSlice.actions.addBooks([
          { bookId: "4", title: "Goosebumps: Night of the Living Dummy" },
          { bookId: "5", title: "Goosebumps: Welcome to Dead House" },
        ]).payload
      ).toMatchSnapshot();
      expect(
        booksSlice.actions.setBook({
          bookId: "3",
          title: "Goosebumps: Night of the Living Dummy",
        }).payload
      ).toMatchSnapshot();
      expect(
        booksSlice.actions.setBooks([
          { bookId: "4", title: "Goosebumps: Night of the Living Dummy" },
          { bookId: "5", title: "Goosebumps: Welcome to Dead House" },
        ]).payload
      ).toMatchSnapshot();
      expect(
        booksSlice.actions.setAllBooks([
          { bookId: "4", title: "Goosebumps: Night of the Living Dummy" },
          { bookId: "5", title: "Goosebumps: Welcome to Dead House" },
        ]).payload
      ).toMatchSnapshot();
      expect(booksSlice.actions.removeBook("2").payload).toMatchSnapshot();
      expect(
        booksSlice.actions.removeBooks(["1", "3"]).payload
      ).toMatchSnapshot();
      expect(booksSlice.actions.removeAllBooks().payload).toMatchSnapshot();
      expect(
        booksSlice.actions.updateBook({
          id: "3",
          changes: { title: "Goosebumps: Night of the Living Dummy" },
        }).payload
      ).toMatchSnapshot();
      expect(
        booksSlice.actions.updateBooks([
          {
            id: "3",
            changes: { title: "Goosebumps: Night of the Living Dummy" },
          },
          { id: "4", changes: { title: "Goosebumps: Welcome to Dead House" } },
        ]).payload
      ).toMatchSnapshot();
      expect(
        booksSlice.actions.upsertBook({
          bookId: "3",
          title: "Goosebumps: Night of the Living Dummy",
        }).payload
      ).toMatchSnapshot();
      expect(
        booksSlice.actions.upsertBooks([
          { bookId: "3", title: "Goosebumps: Night of the Living Dummy" },
          { bookId: "3", title: "Goosebumps: Welcome to Dead House" },
        ]).payload
      ).toMatchSnapshot();
    });
  });

  describe("createNotableEntities", () => {
    it("creates actions and reducers for notable entities", () => {
      const bookReducers = entityAdapterToReducers(booksAdapter, "book");
      const notableBooks = createNotableEntities(booksAdapter, "book", {
        best: ({ id }: { id: string; reason: string }) => id,
        worst: (id: string) => id,
        favorite: (id: string) => id,
      });

      const booksSlice = createSlice("books", {
        ...booksAdapter.getInitialState(),
        ...notableBooks.getInitialState(),
      }).setReducer((builder) =>
        builder.addCases(bookReducers).addCases(notableBooks.reducers)
      );

      const { selectBestBook, selectWorstBook, selectFavoriteBook } =
        notableBooks.getSelectors(booksSlice.selectBooksState);
      const store = configureStore({
        reducer: combineReducers({
          books: booksSlice.reducer,
        }),
      });

      const booksSelectors = booksAdapter.getSelectors(
        booksSlice.selectBooksState
      );

      expect(booksSelectors.selectAll(store.getState())).toMatchSnapshot();
      expect(selectBestBook(store.getState())).toMatchSnapshot(
        "selectBestBook1"
      );
      expect(selectWorstBook(store.getState())).toMatchSnapshot(
        "selectWorstBook1"
      );
      expect(selectFavoriteBook(store.getState())).toMatchSnapshot(
        "selectFavoriteBook1"
      );

      const denoteBestBookAction = booksSlice.actions.denoteBestBook({
        id: "1",
        reason: "It's awesome!",
      });
      const denoteWorstBookAction = booksSlice.actions.denoteWorstBook("2");
      const denoteFavoriteBookAction =
        booksSlice.actions.denoteFavoriteBook("3");

      expect(denoteBestBookAction.payload).toMatchSnapshot();
      expect(denoteWorstBookAction.payload).toMatchSnapshot();
      expect(denoteFavoriteBookAction.payload).toMatchSnapshot();

      store.dispatch(denoteBestBookAction);
      store.dispatch(denoteWorstBookAction);
      store.dispatch(denoteFavoriteBookAction);

      store.dispatch(
        booksSlice.actions.addBooks([
          { bookId: "1", title: "Goosebumps: Monster Blood" },
          { bookId: "2", title: "Goosebumps: Night of the Living Dummy" },
          { bookId: "3", title: "Goosebumps: Welcome to Dead House" },
        ])
      );

      expect(selectBestBook(store.getState())).toMatchSnapshot(
        "selectBestBook2"
      );
      expect(selectWorstBook(store.getState())).toMatchSnapshot(
        "selectWorstBook2"
      );
      expect(selectFavoriteBook(store.getState())).toMatchSnapshot(
        "selectFavoriteBook2"
      );
    });
  });
});
