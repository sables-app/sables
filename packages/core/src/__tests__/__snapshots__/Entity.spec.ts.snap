// Vitest Snapshot v1

exports[`Entity > createNotableEntities > creates actions and reducers for notable entities > selectBestBook1 1`] = `undefined`;

exports[`Entity > createNotableEntities > creates actions and reducers for notable entities > selectBestBook2 1`] = `
{
  "bookId": "1",
  "title": "Goosebumps: Monster Blood",
}
`;

exports[`Entity > createNotableEntities > creates actions and reducers for notable entities > selectFavoriteBook1 1`] = `undefined`;

exports[`Entity > createNotableEntities > creates actions and reducers for notable entities > selectFavoriteBook2 1`] = `
{
  "bookId": "3",
  "title": "Goosebumps: Welcome to Dead House",
}
`;

exports[`Entity > createNotableEntities > creates actions and reducers for notable entities > selectWorstBook1 1`] = `undefined`;

exports[`Entity > createNotableEntities > creates actions and reducers for notable entities > selectWorstBook2 1`] = `
{
  "bookId": "2",
  "title": "Goosebumps: Night of the Living Dummy",
}
`;

exports[`Entity > createNotableEntities > creates actions and reducers for notable entities 1`] = `[]`;

exports[`Entity > createNotableEntities > creates actions and reducers for notable entities 2`] = `
{
  "id": "1",
  "reason": "It's awesome!",
}
`;

exports[`Entity > createNotableEntities > creates actions and reducers for notable entities 3`] = `"2"`;

exports[`Entity > createNotableEntities > creates actions and reducers for notable entities 4`] = `"3"`;

exports[`Entity > entityAdapterToReducers > renames actions and reducers from entity adapters 1`] = `
{
  "bookId": "3",
  "title": "Goosebumps: Night of the Living Dummy",
}
`;

exports[`Entity > entityAdapterToReducers > renames actions and reducers from entity adapters 2`] = `
[
  {
    "bookId": "4",
    "title": "Goosebumps: Night of the Living Dummy",
  },
  {
    "bookId": "5",
    "title": "Goosebumps: Welcome to Dead House",
  },
]
`;

exports[`Entity > entityAdapterToReducers > renames actions and reducers from entity adapters 3`] = `
{
  "bookId": "3",
  "title": "Goosebumps: Night of the Living Dummy",
}
`;

exports[`Entity > entityAdapterToReducers > renames actions and reducers from entity adapters 4`] = `
[
  {
    "bookId": "4",
    "title": "Goosebumps: Night of the Living Dummy",
  },
  {
    "bookId": "5",
    "title": "Goosebumps: Welcome to Dead House",
  },
]
`;

exports[`Entity > entityAdapterToReducers > renames actions and reducers from entity adapters 5`] = `
[
  {
    "bookId": "4",
    "title": "Goosebumps: Night of the Living Dummy",
  },
  {
    "bookId": "5",
    "title": "Goosebumps: Welcome to Dead House",
  },
]
`;

exports[`Entity > entityAdapterToReducers > renames actions and reducers from entity adapters 6`] = `"2"`;

exports[`Entity > entityAdapterToReducers > renames actions and reducers from entity adapters 7`] = `
[
  "1",
  "3",
]
`;

exports[`Entity > entityAdapterToReducers > renames actions and reducers from entity adapters 8`] = `undefined`;

exports[`Entity > entityAdapterToReducers > renames actions and reducers from entity adapters 9`] = `
{
  "changes": {
    "title": "Goosebumps: Night of the Living Dummy",
  },
  "id": "3",
}
`;

exports[`Entity > entityAdapterToReducers > renames actions and reducers from entity adapters 10`] = `
[
  {
    "changes": {
      "title": "Goosebumps: Night of the Living Dummy",
    },
    "id": "3",
  },
  {
    "changes": {
      "title": "Goosebumps: Welcome to Dead House",
    },
    "id": "4",
  },
]
`;

exports[`Entity > entityAdapterToReducers > renames actions and reducers from entity adapters 11`] = `
{
  "bookId": "3",
  "title": "Goosebumps: Night of the Living Dummy",
}
`;

exports[`Entity > entityAdapterToReducers > renames actions and reducers from entity adapters 12`] = `
[
  {
    "bookId": "3",
    "title": "Goosebumps: Night of the Living Dummy",
  },
  {
    "bookId": "3",
    "title": "Goosebumps: Welcome to Dead House",
  },
]
`;

exports[`Entity > entityAdapterToReducers > slice dependency integration 1`] = `
{
  "entities": {
    "1": {
      "bookId": "1",
      "title": "Harry Potter",
    },
  },
  "ids": [
    "1",
  ],
}
`;

exports[`Entity > entityAdapterToReducers > slice dependency integration 2`] = `
{
  "entities": {
    "1": {
      "bookId": "1",
      "title": "Harry Potter",
    },
  },
  "ids": [
    "1",
  ],
}
`;

exports[`Entity > entityAdapterToReducers > slice dependency integration 3`] = `
[
  {
    "bookId": "1",
    "title": "Harry Potter",
  },
]
`;
